Fix1008 sony2010e

clanmills opened the pull request:
@sridharb1 and I are collaborating on this.  For the moment, it's Work-in-progress.

clanmills committed:
Fix printStructure to handle Sony1 MakerNote (with header SONY DSC)

clanmills committed:
Sridhar's initial 2010e support.

clanmills committed:
Add tests

clanmills committed:
Changes3.  Will next investigate FAIL: test_run (bugfixes.github.test_issue_981.CanonAfInfoTest)

D4N commented:
Could you please fix the diffs for some of the `_int.cpp` files? It is quite impossible to find out what you changed, because the whole file got mangled.

sridharb1 committed:
Needs winsock2.h to compile

clanmills commented:
It's been mangled by the line-endings.

The code provided by @sridharb1 has Windows CR/LF line endings which SmartGit hid from me.  I think I can put this right after @sridharb1 has finished what he's doing at the moment.

SVN has a file property to say "This file is to use EOL: NL" to prevent this kind of clobbering.  Do you know if there's something similar in Git?

clanmills commented:
```
562 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ git diff -b fix1008_Sony2010e 0.27-maintenance > ~/temp/dc5523f6.diff
563 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ (cd ~/temp ; tar czf dc5523f6.tar.gz dc5523f6.diff) 
564 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ ls -l ~/temp/dc5523f6.tar.gz 
-rw-r--r--  1 rmills  staff  9397 30 Sep 12:06 /Users/rmills/temp/dc5523f6.tar.gz
565 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ 
```
[dc5523f6.tar.gz](https://github.com/Exiv2/exiv2/files/3670556/dc5523f6.tar.gz)

sridharb1 committed:
Need to update to new lens definitions

sridharb1 committed:
Updated to new CanonLenses definitions

sridharb1 committed:
Catchup to Canon cameras and lenses from exiftool

D4N commented:
Robin Mills <notifications@github.com> writes:

> It's been mangled by the line-endings.
>
> The code provided by @sridharb1 has Windows CR/LF line endings which SmartGit hid from me.  I think I can put this right after @sridharb1 has finished what he's doing at the moment.
>
> SVN has a file property to say "This file is to use EOL: NL" to prevent this kind of clobbering.  Do you know if there's something similar in Git? 
>

https://help.github.com/en/articles/configuring-git-to-handle-line-endings

> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/Exiv2/exiv2/pull/1009#issuecomment-536508925

sridharb1 commented:
I am done with pushing it into fix1008...
Line endings should not be a problem. They should be translated automatically by git to the line ending you use locally.
BTW, the line endings in the test folder for the .sh and functions.source is wrong. They are checked in as text files and thus come with a crlf for me, but it needs just lf as I am executing them in a bash shell. I deal with this locally after every update.

    On Monday, 30 September, 2019, 4:20:42 pm IST, Robin Mills <notifications@github.com> wrote:  
 
 
It's been mangled by the line-endings.

The code provided by @sridharb1 has Windows CR/LF line endings which SmartGit hid from me. I think I can put this right after @sridharb1 has finished what he's doing at the moment.

SVN has a file property to say "This file is to use EOL: NL" to prevent this kind of clobbering. Do you know if there's something similar in Git?


—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

sridharb1 commented:
This is not the problem. core.autocrlf is true for me.
This means that it should get converted to lf upon checkin.

   On Monday, 30 September, 2019, 4:45:18 pm IST, D4N <notifications@github.com> wrote:  
 
 Robin Mills <notifications@github.com> writes:

> It's been mangled by the line-endings.
>
> The code provided by @sridharb1 has Windows CR/LF line endings which SmartGit hid from me. I think I can put this right after @sridharb1 has finished what he's doing at the moment.
>
> SVN has a file property to say "This file is to use EOL: NL" to prevent this kind of clobbering. Do you know if there's something similar in Git? 
>

https://help.github.com/en/articles/configuring-git-to-handle-line-endings

> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/Exiv2/exiv2/pull/1009#issuecomment-536508925


—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

D4N commented on xmpsdk/src/XMPUtils.cpp:
```diff
@@ -5,7 +5,7 @@
 // of the Adobe license agreement accompanying it.
 // =================================================================================================
-
+#define NOMINMAX
```
Huh, why is this suddenly necessary? Also, we shouldn't touch the xmpsdk's source unless absolutely required.

clanmills replied:
@sridharb1 replied _On, windows there is a problem if you don't have it._

D4N replied:
Then this define should be added via cmake.

piponazo replied:
The CMake code is in fact already including that definition. Take a look at `cmake/compilerFlags.cmake`

sridharb1 commented:
On, windows there is a problem if you don't have it.

    On Monday, 30 September, 2019, 6:44:21 pm IST, D4N <notifications@github.com> wrote:  
 
 
@D4N commented on this pull request.

In xmpsdk/src/XMPUtils.cpp:
> @@ -5,7 +5,7 @@
 // NOTICE:	Adobe permits you to use, modify, and distribute this file in accordance with the terms
 // of the Adobe license agreement accompanying it.
 // =================================================================================================
-
+#define NOMINMAX

Huh, why is this suddenly necessary? Also, we shouldn't touch the xmpsdk's source unless absolutely required.


—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

clanmills committed:
Fix CR/LF Endings.  Added include <math.h> to compile pow() on MacOS-X.

clanmills commented:
@sridharb1  I've "tweaked" the line endings (with dos2unix) and that's much better.  A couple of matters to be resolved:

1) There are a great many changes in canonmn_int.cpp from the original.

My impression is that you generated that file with a script that reads code in ExifTool.  It's added/modified new/changed lens definitions.  It has also modified unchanged definitions.  And while this is cosmetic, if it's possible to avoid changing existing code, this is desirable.

2) FAIL: test_run (bugfixes.github.test_issue_981.CanonAfInfoTest)

Because this branch is from 0.27-maintenance, you don't have my fix for #981.  It's on a branch waiting to be approved/merged into 0.27-maintenance.  I've invited @derselbst  to approve the changes for #981 (and #1001 and #1010).  I would expect you to have both the code AND the test harness changes (or neither).  Let's get everything up-to-date and we'll fix it from there.  I'll deal with this and you can ignore this at the moment.

clanmills committed:
Modify test_issue_981.py to pass (to enable CI to pass).

clanmills commented:
I've removed WIP from title.  This PR is ready for review.

I've modified test_issue_981.CanonAfInfoTest to pass and this PR should pass the CI.  I will be dealing with #981 (and his descendants) for v0.27.3.  I'm puzzled that test code for #981 is on branch 0.27-maintenance.

sridharb1 commented:
This is by design. It is best if canonmn_int.cpp is in sync with the naming conventions of exiftool (which I consider the reference standard) and not other names that was added in an adhoc manner over the years. This would also improve future maintability to add just the differences and not search through the mountain of spurious differences to arrive at the meaningful changes.
Regards,Sridhar



    On Monday, 30 September, 2019, 8:34:12 pm IST, Robin Mills <notifications@github.com> wrote:  
 
 
@sridharb1 I've "tweaked" the line endings (with dos2unix) and that's much better. A couple of matters to be resolved:
   
   - There are a great many changes in canonmn_int.cpp from the original.

My impression is that you generated that file with a script that reads code in ExifTool. It's added/modified new/changed lens definitions. It has also modified unchanged definitions. And while this is cosmetic, if it's possible to avoid changing existing code, this is desirable.
   
   - FAIL: test_run (bugfixes.github.test_issue_981.CanonAfInfoTest)

Because this branch is from 0.27-maintenance, you don't have my fix for #981. It's on a branch waiting to be approved/merged into 0.27-maintenance. I've invited @derselbst to approve the changes for #981 (and #1001 and #1010). I would expect you to have both the code AND the test harness changes (or neither). Let's get everything up-to-date and we'll fix it from there. I'll deal with this and you can ignore this at the moment.


—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

clanmills commented:
Thanks for the explanation.  If you have a script to generate this code, could you share it with us please.

I agree that ExifTool is the "reference standard".  Neils worked on Exiv2 for 3 or 4 years and made many changes concerning Exiv2/Exiftool compatibility.  As I was quite new to the project (and dealing with cross-platform build/test), I didn't really know anything about his work.  

Since Neils left the project (to return to college), I introduced the ~/.exiv2 file to enable users to define lenses.  @nkbj You're welcome back whenever you'd like to join us.

tester0077 commented:
To avoid changing the XMPSDK code, perhaps it should go into the Windows 
CMake code

Arnold

On 2019-09-30 9:44 AM, Robin Mills wrote:
>
> *@clanmills* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In xmpsdk/src/XMPUtils.cpp 
> <https://github.com/Exiv2/exiv2/pull/1009#discussion_r329678196>:
>
> > @@ -5,7 +5,7 @@
>   // NOTICE:	Adobe permits you to use, modify, and distribute this file in accordance with the terms
>   // of the Adobe license agreement accompanying it.
>   // =================================================================================================
> -
> +#define NOMINMAX
>
> @sridharb1 <https://github.com/sridharb1> replied /On, windows there 
> is a problem if you don't have it./
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/Exiv2/exiv2/pull/1009?email_source=notifications&email_token=ACFCLPEEQCCHJMFOGUGAEYLQMIUGJA5CNFSM4I3RJJFKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCGLP7BA#discussion_r329678196>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ACFCLPCEVBWJKFE2MIICIOTQMIUGJANCNFSM4I3RJJFA>.
>
-- 
Fight Spam - report it with wxSR 0.8
Vista & Win 7, 8 & 10 compatible
http://www.columbinehoney.net/wxSR.shtml

clanmills commented:
Interesting idea @tester0077 I believe @sridharb1 is using his own vcxproj/sln files (derived from the now removed msvc/ build environment), so he's not using CMake.

The principle that this can be defined in the project (and not explicitly in the code) sounds good.  It's rather strange that this seems to be required at all, as I don't require this and we've built Exiv2 with all versions of Visual Studio from 2003 onwards.  

On the other hand, this is a minor matter and if it's needed to build the code, so be it.  @D4N is correct to say that Team Exiv2 endeavours to avoid changing xmp/ code.

sridharb1 commented:
> 
> 
> Thanks for the explanation. If you have a script to generate this code, could you share it with us please.
> 
> I agree that ExifTool is the "reference standard". Neils worked on Exiv2 for 3 or 4 years and made many changes concerning Exiv2/Exiftool compatibility. As I was quite new to the project (and dealing with cross-platform build/test), I didn't really know anything about his work.
> 
> Since Neils left the project (to return to college), I introduced the ~/.exiv2 file to enable users to define lenses. @nkbj You're welcome back whenever you'd like to join us.

I use an excel macro to generate the lens and camera list. This is not very sophisticated, as in, it needs manual input and cleanup afterwards, but it does ~99% of the work which saves a ton of time that would be spent in otherwise changing the format from exiftool/perl to exiv2/c. If interested, I can share it. This may be used to get other "big" systems like Nikon in sync, but I didn't venture there because of other priorities.

sridharb1 commented:
This is strange. Regarding dos2unix, this is not the correct practice to do dos2unix before checking things in. Modern scm systems should handle these things transparently. If the repository thinks that this file should always have unix line endings (even on Windows), then the file is incorrectly checked in.

>     1. FAIL: test_run (bugfixes.github.test_issue_981.CanonAfInfoTest)
> 
> 
> Because this branch is from 0.27-maintenance, you don't have my fix for #981. It's on a branch waiting to be approved/merged into 0.27-maintenance. I've invited @derselbst to approve the changes for #981 (and #1001 and #1010). I would expect you to have both the code AND the test harness changes (or neither). Let's get everything up-to-date and we'll fix it from there. I'll deal with this and you can ignore this at the moment.

I think it is a mystery to even you looking at your other comment.

clanmills commented:
There are many mysteries in life that I can't explain.  For sure the popularity of **git** is one of them.

Anyway, this PR is looking much better and almost passing the CI.  There's at least one more "mystery" and that's CentOS is failing to build and reports:
```
/usr/bin/bash: line 97: python3: command not found
```
It's being doing this for months and blocking everything.  However, @D4N fixed that yesterday on the 0.27-maintenance branch and @piponazo approved/merged it and that's a welcome step forward.

Next mystery is how to get the outstanding PRs (including this) approved and merged onto 0.27-maintenance.  After that, I still have about 3 issues to fix and then Exiv2 v0.27.3-RC1 can be published.

sridharb1 commented:
I have the answer for why the line endings are messed up. You run a Mac-based environment and I use Windows. When you extracted the files from the zip file, the files ended up being crlf in an lf world. When you checked them in, git thought it doesn't need to translate anything (because the source is an lf world) and thus, they ended up in the repository with crlf and caused problems thereafter.

clanmills commented:
Yes.  I thought I explained that yesterday!  It impacted the 3 files which you send in the zip.

I use SmartGit as my GUI on Git (although I also use the command-line git) and it's strange that informed me of the changes I made in your code for the selector() and `#include <math.h>` and yet it was silent about line-endings.

Anyway, the sun's shining this morning and all is well with this PR, except the CentOS/python3 issue.

sridharb1 reviewed (dismissed):
There are fixes for 981, 1004 and 1009 in the same branch. Given that those have their own branches, these changes are not necessary to be included here.

Also, in case we run into this line-ending problem again, could we see if the version that was incorrectly checked in was backed out instead of adding a new version with dos2unix or unix2dos done on it? That would allow the history to be accurate. I don't mind if the history captures that a commit was made and then subsequently deleted. I don't want to see a diff showing the whole file has changed.

clanmills commented:
I don't have the git skills to back out something.

D4N reviewed (changes requested):
There is still files with a huge diff left, please fix that.

Also, I am not sure if this belongs into a maintenance branch to be honest, as it changes the output of exiv2 by quite a bit.

D4N commented on src/tiffimage_int.cpp:
```diff
@@ -749,7 +749,7 @@ namespace Exiv2 {
     //! Sony1 Camera Settings binary array - configuration
     extern const ArrayCfg sony1CsCfg = {
         sony1CsId,        // Group for the elements
-        bigEndian,        // Big endian
```
Why the sudden whitespace changes now?

clanmills replied:
What whitespace changes?  In the code, in the output?

sridharb1 replied:
Whitespace changes were done by VS, regrettably. It changed the spaces to tabs. I have changed the setting that prevents this from happening again.

D4N commented on tests/bugfixes/github/test_issue_981.py:
```diff
@@ -6,59 +6,12 @@ class CanonAfInfoTest(metaclass=CaseMeta):
-                 "$exiv2 -pv --grep Points   $filenameC",
-                 "$exiv2 -pt --grep Points   $filenameC",
-                 "$exiv2 -pv --grep Primary  $filenameD",
-                 "$exiv2 -pt --grep Primary  $filenameD",
```
Why are these tests dropped?

clanmills replied:
I had to remove them to get the test to pass.  I will have to restore them when all the open PRs are eventually merged.  I'm totally frustrated by having about 12 different PRs open and unreviewed.

D4N commented on src/canonmn_int.cpp:
```diff
@@ -741,369 +763,457 @@ namespace Exiv2 {
 
     //! LensType, tag 0x0016
     extern const TagDetails canonCsLensType[] = {
```
The diff of the following array looks like everything got changed, but when glancing over it I have a hard time to actually spot the changes. What happened here?

sridharb1 replied:
The original code was pretty-formatted to align with lenses that have 4 spaces. Unfortunately, there came along lens #10001 and thus not aligned. I just gave ourselves more breathing room by using 5 spaces throughout.

clanmills commented:
I don't know what to do about the dos2unix matter.  I can think of two possibilities:

1) Replace this PR
Find the resulting patch implied by this PR (with git diff)
Open a new branch and apply the patch
Close this PR
Review/approve/merge the new branch

2) I give up and retire.  I am exhausted, frustrated and totally browned off.

sridharb1 commented:
> 
> 
> There is still files with a huge diff left, please fix that.
> 
> Also, I am not sure if this belongs into a maintenance branch to be honest, as it changes the output of exiv2 by quite a bit.

Yes, it aligns exiv2 and exiftool, at least for Canon lenses and cameras. There may be people who look for the old output, they will need to update their scripts. The advantages of aligning with exiftool, I believe, outweighs the potential incidents that may arise, even with a maintenance release patch.

D4N commented:
Luis Díaz Más <notifications@github.com> writes:

> piponazo commented on this pull request.
>
>
>
>> @@ -5,7 +5,7 @@
>  // NOTICE:	Adobe permits you to use, modify, and distribute this file in accordance with the terms
>  // of the Adobe license agreement accompanying it.
>  // =================================================================================================
> -
> +#define NOMINMAX
>
> The CMake code is in fact already including that definition. Take a look at `cmake/compilerFlags.cmake`
>

I've dropped the respective commit now.

sridharb1 commented:
Ok


Sent from Yahoo Mail for iPhone


On Friday, October 4, 2019, 20:42, D4N <notifications@github.com> wrote:

Luis Díaz Más <notifications@github.com> writes:

> piponazo commented on this pull request.
>
>
>
>> @@ -5,7 +5,7 @@
> // NOTICE: Adobe permits you to use, modify, and distribute this file in accordance with the terms
> // of the Adobe license agreement accompanying it.
> // =================================================================================================
> -
> +#define NOMINMAX
>
> The CMake code is in fact already including that definition. Take a look at `cmake/compilerFlags.cmake`
>

I've dropped the respective commit now.


—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

clanmills commented:
@sridharb1 I've totally lost the plot on this.  This PR is conflicted with the current state of branch `0.27-maintenance` due to changes in `src/canonmn_int.cpp`.  It is also likely that this PR (and associate branch  `fix1008_Sony2010e `) could "bloat" the repository with the file with the CRLF conflicts.

Can we start again, please?  Can you open a new issue with the changes and I will submit a new PR.  I am closing this PR.  I would also like to delete the branch `fix1008_Sony2010e`

sridharb1 commented:
Ok, we will start again. I will hunt down those fixes for sony2010e (remember that there is 2010a-f, I think and I implemented only 2010e because I had an NEX-6 that identified itself as this subset). I say this to just note that the catching up to exiftool is significantly inadequate. But, at least, we are better off than before.

I am going to make these fixes in my fork of 0.27-maintenance. I am hoping that you can pull these into the main branch.

sridharb1 commented:
I have pushed my sony2010e changes to my 0.27-maintenance branch. Could you please pick them up?

clanmills commented:
@sridharb1 I don't have the git skills to know how to "pick them up".  I believe it's possible for you to submit a PR from https://github.com/sridharb1/exiv2 to https://github.com/exiv2/exiv2

Team: Could one of you help @sridharb1 to get this done, please? 

This card keeps me out of git hell and is my my total knowledge of git:
![GitIdiotCard](https://user-images.githubusercontent.com/529982/77848618-f1572800-71bd-11ea-84ae-4d799c55e76c.jpg)

sridharb1 commented:
[As a preface, it is best to note that my git skills are pretty limited as well.]

I think it is best if you keep this PR closed as this conflates two changes that I did during that period.

One was the sony2010e, which we are trying to merge now. The other was the canon lens and camera sync with exiftool. The latter caused the unfortunate line-ending issues. I would keep the latter separate.

I tried merging my fork with the base, but unfortunately, it pulls in some changes that I had to do to get CMake past some hurdles in my machine, which are probably inappropriate for distribution. Realistically, these are the types of problems that I face with CMake, but they will not be acknowledged because some platforms and ways of doing things are irrelevant.

So, I will create a branch that has only those (sony2010e) and create a clean pull request for you.

clanmills commented:
OK.  I'm going to close this and delete the branch from GitHub.

I'm sorry to hear that CMake is causing you distress.  How about you get your code changes ready on your machine and we work together using your `contrib/msvc` build environment.

Please get a clean/fresh copy of the code.  Don't mix your current ball of stuff with ours.
I suggest that you work in the repos https://github.com/exiv2/exiv2 and sync up to the latest 0.27-maintenance code which includes `contrib/msvc`.  Then create a branch on your machine.

```
$ cd ..../where/you/work
$ git clone https://github.com/exiv2/exiv2 --branch 0.27-maintenance 0.27-maintenance --depth 1
$ cd 0.27-maintenance
$ git fetch --unshallow
$ git checkout -b sridharb1_sony2010e_0.27   # or CanonExifToolLensChanges (or something)
```

Build it with contrib/msvc/exiv2.sln.  Then we'll talk again.

Let's start with the smallest/simplest patch you'd like to prepare.  Together, we can do this.

------------------

For my own information, I've looked at the changes in #1009:
```
$ curl https://patch-diff.githubusercontent.com/raw/Exiv2/exiv2/pull/1009.patch > ~/temp/1009.patch
583 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ git apply ~/temp/1009.patch 
/Users/rmills/temp/1009.patch:99: trailing whitespace.
	
/Users/rmills/temp/1009.patch:2348: trailing whitespace.
// ***************************************************************** -*- C++ -*-
/Users/rmills/temp/1009.patch:2349: trailing whitespace.
/*
/Users/rmills/temp/1009.patch:2350: trailing whitespace.
 * Copyright (C) 2004-2018 Exiv2 authors
/Users/rmills/temp/1009.patch:2351: trailing whitespace.
 * This program is part of the Exiv2 distribution.
error: patch failed: test/data/exiv2-test.out:1890
error: test/data/exiv2-test.out: patch does not apply
error: patch failed: tests/bugfixes/redmine/test_issue_1252.py:16
error: tests/bugfixes/redmine/test_issue_1252.py: patch does not apply
error: patch failed: src/canonmn_int.cpp:118
error: src/canonmn_int.cpp: patch does not apply
error: patch failed: tests/bugfixes/github/test_issue_981.py:6
error: tests/bugfixes/github/test_issue_981.py: patch does not apply
584 rmills@rmillsmbp:~/gnu/github/exiv2/0.27-maintenance $ 
```
Inside the patch, there are changes to 20 files.
```
$ grep -e "--- a/" ~/temp/1009.patch 
--- a/src/image.cpp
--- a/src/makernote_int.cpp
...
--- a/tests/bugfixes/github/test_issue_981.py
$
```

sridharb1 commented:
That's fine.

However, there is no need to mix two things here. My changes for compiling with VS can be kept separate from my feature of supporting the sony2010 tag.

I am currently working on getting you a clean pull request for sony2010.

I will look into the VS project files afterward.

sridharb1 commented:
I have done this now. Please see #1126 and bring it into 0.27-maintenance. This does have two small fixes to image.cpp that are good for the distribution as well.
